File: src/xbe.c
#include "xbe.h"
#include "kernel_exports.h"

#include <stdio.h>
#include <string.h>

#define ENTRY_RETAIL_XOR 0xA8FC57ABu
#define ENTRY_DEBUG_XOR  0x94859D4Bu
#define THUNK_RETAIL_XOR 0x5B6D40B6u
#define THUNK_DEBUG_XOR  0xEFB1F152u

static uint32_t rd_u32(const uint8_t* p, size_t len, size_t off) {
    if (off + 4 > len) return 0;
    return (uint32_t)p[off] | ((uint32_t)p[off + 1] << 8) |
           ((uint32_t)p[off + 2] << 16) | ((uint32_t)p[off + 3] << 24);
}

int xbe_parse(const uint8_t* data, size_t len, Xbe* out) {
    memset(out, 0, sizeof *out);
    if (len < 0x180 || memcmp(data, "XBEH", 4) != 0) {
        fprintf(stderr, "not an XBE (bad 'XBEH' magic)\n");
        return 1;
    }

    out->raw = malloc(len);
    memcpy(out->raw, data, len);
    out->raw_len = len;
    const uint8_t* p = out->raw;

    uint32_t base_addr    = rd_u32(p, len, 0x0104);
    uint32_t headers_size = rd_u32(p, len, 0x0108);
    uint32_t image_size   = rd_u32(p, len, 0x010C);
    uint32_t num_sections = rd_u32(p, len, 0x011C);
    uint32_t sec_hdr_va   = rd_u32(p, len, 0x0120);
    uint32_t entry_raw    = rd_u32(p, len, 0x0128);
    uint32_t tls_va       = rd_u32(p, len, 0x012C);
    uint32_t thunk_raw    = rd_u32(p, len, 0x0158);

    uint32_t entry_retail = entry_raw ^ ENTRY_RETAIL_XOR;
    uint32_t entry_debug  = entry_raw ^ ENTRY_DEBUG_XOR;
    int is_debug; uint32_t entry;
    if (entry_retail >= base_addr && entry_retail < base_addr + image_size) {
        entry = entry_retail; is_debug = 0;
    } else if (entry_debug >= base_addr && entry_debug < base_addr + image_size) {
        entry = entry_debug; is_debug = 1;
    } else {
        entry = entry_retail; is_debug = 0;
    }
    uint32_t thunk = is_debug ? (thunk_raw ^ THUNK_DEBUG_XOR) : (thunk_raw ^ THUNK_RETAIL_XOR);

    out->base_address      = base_addr;
    out->image_size        = image_size;
    out->headers_size      = headers_size;
    out->entry_point       = entry;
    out->kernel_thunk_addr = thunk;
    out->tls_addr          = tls_va;
    out->is_debug          = is_debug;

    size_t sec_base = (size_t)(sec_hdr_va - base_addr);
    for (uint32_t i = 0; i < num_sections; i++) {
        size_t o = sec_base + (size_t)i * 56;
        if (o + 56 > len) break;
        XbeSection s;
        memset(&s, 0, sizeof s);
        s.flags        = rd_u32(p, len, o + 0);
        s.virtual_addr = rd_u32(p, len, o + 4);
        s.virtual_size = rd_u32(p, len, o + 8);
        s.raw_addr     = rd_u32(p, len, o + 12);
        s.raw_size     = rd_u32(p, len, o + 16);
        uint32_t name_va = rd_u32(p, len, o + 20);
        if (name_va != 0) {
            size_t no = (size_t)(name_va - base_addr);
            if (no < len) {
                size_t end = no;
                while (end < len && end < no + 32 && p[end] != 0) end++;
                size_t n = end - no;
                if (n > sizeof(s.name) - 1) n = sizeof(s.name) - 1;
                memcpy(s.name, p + no, n);
                s.name[n] = 0;
            }
        }
        vec_push(&out->sections, s);
    }

    /* kernel import thunk table: u32 array until 0; high bit => ordinal */
    if (thunk != 0) {
        long file_off = -1;
        for (size_t i = 0; i < out->sections.len; i++) {
            XbeSection* s = &out->sections.data[i];
            if (thunk >= s->virtual_addr && thunk < xbe_sec_end(s)) {
                file_off = (long)(s->raw_addr + (thunk - s->virtual_addr));
                break;
            }
        }
        if (file_off < 0) file_off = (long)thunk;
        for (size_t off = (size_t)file_off; off + 4 <= len; off += 4) {
            uint32_t val = rd_u32(p, len, off);
            if (val == 0) break;
            if (val & 0x80000000u) {
                XbeKernelImport imp;
                memset(&imp, 0, sizeof imp);
                imp.ordinal = (int)(val & 0x7FFFFFFF);
                xbox_kernel_export_name_buf(imp.ordinal, imp.name, sizeof imp.name);
                imp.thunk_addr = (uint32_t)(thunk + (off - (size_t)file_off));
                vec_push(&out->kernel_imports, imp);
            }
        }
    }
    return 0;
}

int xbe_load(const char* path, Xbe* out) {
    FILE* f = fopen(path, "rb");
    if (!f) { fprintf(stderr, "cannot open %s\n", path); return 1; }
    fseek(f, 0, SEEK_END);
    long n = ftell(f);
    fseek(f, 0, SEEK_SET);
    if (n <= 0) { fclose(f); fprintf(stderr, "empty file %s\n", path); return 1; }
    uint8_t* buf = malloc((size_t)n);
    size_t got = fread(buf, 1, (size_t)n, f);
    fclose(f);
    if (got != (size_t)n) { free(buf); fprintf(stderr, "short read %s\n", path); return 1; }
    int rc = xbe_parse(buf, (size_t)n, out);
    free(buf);
    return rc;
}

void xbe_free(Xbe* x) {
    free(x->raw);
    vec_free(&x->sections);
    vec_free(&x->kernel_imports);
    memset(x, 0, sizeof *x);
}

File: src/OgXbox.Recomp.Codegen/Binary/Xbe.cs
// Minimal XBE (Xbox Executable) parser — header, section table, kernel imports.
//
// Independent implementation from the XBE format (offsets cross-checked against
// the X-Men recomp's tools/xbe_parser). Only what the recompiler needs.

using System;
using System.Buffers.Binary;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace OgXbox.Recomp.Codegen.Binary;

[Flags]
public enum XbeSectionFlags : uint
{
    Writable = 0x01,
    Preload = 0x02,
    Executable = 0x04,
    InsertedFile = 0x08,
    HeadPageReadOnly = 0x10,
    TailPageReadOnly = 0x20,
}

public sealed class XbeSection
{
    public required string Name { get; init; }
    public required uint VirtualAddress { get; init; }
    public required uint VirtualSize { get; init; }
    public required uint RawAddress { get; init; }
    public required uint RawSize { get; init; }
    public required XbeSectionFlags Flags { get; init; }

    public bool IsExecutable => Flags.HasFlag(XbeSectionFlags.Executable);
    public bool IsWritable => Flags.HasFlag(XbeSectionFlags.Writable);
    public uint VirtualEnd => VirtualAddress + VirtualSize;
}

public sealed class XbeKernelImport
{
    public required int Ordinal { get; init; }
    public required string Name { get; init; }
    public required uint ThunkAddress { get; init; }
}

public sealed class Xbe
{
    private const uint EntryRetailXor = 0xA8FC57AB;
    private const uint EntryDebugXor = 0x94859D4B;
    private const uint ThunkRetailXor = 0x5B6D40B6;
    private const uint ThunkDebugXor = 0xEFB1F152;

    public required byte[] RawData { get; init; }
    public required uint BaseAddress { get; init; }
    public required uint ImageSize { get; init; }
    public required uint HeadersSize { get; init; }
    public required uint EntryPoint { get; init; }
    public required uint KernelThunkAddress { get; init; }
    public required uint TlsAddress { get; init; }
    public required bool IsDebug { get; init; }
    public required IReadOnlyList<XbeSection> Sections { get; init; }
    public required IReadOnlyList<XbeKernelImport> KernelImports { get; init; }

    public static Xbe Load(string path) => Parse(File.ReadAllBytes(path));

    public static Xbe Parse(byte[] data)
    {
        if (data.Length < 0x180 || Encoding.ASCII.GetString(data, 0, 4) != "XBEH")
            throw new InvalidDataException("Not an XBE (bad 'XBEH' magic).");

        uint U32(int off) => BinaryPrimitives.ReadUInt32LittleEndian(data.AsSpan(off, 4));

        uint baseAddr = U32(0x0104);
        uint headersSize = U32(0x0108);
        uint imageSize = U32(0x010C);
        uint numSections = U32(0x011C);
        uint sectionHeadersVa = U32(0x0120);
        uint entryRaw = U32(0x0128);
        uint tlsVa = U32(0x012C);
        uint thunkRaw = U32(0x0158);

        // Header data is mapped at [baseAddr, baseAddr + headersSize) == file [0, headersSize).
        int VaToHeaderOffset(uint va) => (int)(va - baseAddr);

        uint entryRetail = entryRaw ^ EntryRetailXor;
        uint entryDebug = entryRaw ^ EntryDebugXor;
        bool isDebug;
        uint entry;
        if (entryRetail >= baseAddr && entryRetail < baseAddr + imageSize)
        {
            entry = entryRetail; isDebug = false;
        }
        else if (entryDebug >= baseAddr && entryDebug < baseAddr + imageSize)
        {
            entry = entryDebug; isDebug = true;
        }
        else
        {
            entry = entryRetail; isDebug = false;
        }

        uint thunk = isDebug ? (thunkRaw ^ ThunkDebugXor) : (thunkRaw ^ ThunkRetailXor);

        var sections = new List<XbeSection>((int)numSections);
        int secBase = VaToHeaderOffset(sectionHeadersVa);
        for (int i = 0; i < numSections; i++)
        {
            int o = secBase + i * 56;
            var flags = (XbeSectionFlags)U32(o + 0);
            uint vaddr = U32(o + 4);
            uint vsize = U32(o + 8);
            uint raddr = U32(o + 12);
            uint rsize = U32(o + 16);
            uint nameVa = U32(o + 20);

            string name = "";
            if (nameVa != 0)
            {
                int no = VaToHeaderOffset(nameVa);
                if (no >= 0 && no < data.Length)
                {
                    int end = no;
                    while (end < data.Length && end < no + 32 && data[end] != 0) end++;
                    name = Encoding.ASCII.GetString(data, no, end - no);
                }
            }

            sections.Add(new XbeSection
            {
                Name = name,
                VirtualAddress = vaddr,
                VirtualSize = vsize,
                RawAddress = raddr,
                RawSize = rsize,
                Flags = flags,
            });
        }

        var imports = ParseKernelImports(data, sections, thunk);

        return new Xbe
        {
            RawData = data,
            BaseAddress = baseAddr,
            ImageSize = imageSize,
            HeadersSize = headersSize,
            EntryPoint = entry,
            KernelThunkAddress = thunk,
            TlsAddress = tlsVa,
            IsDebug = isDebug,
            Sections = sections,
            KernelImports = imports,
        };
    }

    private static List<XbeKernelImport> ParseKernelImports(
        byte[] data, List<XbeSection> sections, uint thunkVa)
    {
        var imports = new List<XbeKernelImport>();
        if (thunkVa == 0) return imports;

        int? fileOff = null;
        foreach (var s in sections)
        {
            if (thunkVa >= s.VirtualAddress && thunkVa < s.VirtualEnd)
            {
                fileOff = (int)(s.RawAddress + (thunkVa - s.VirtualAddress));
                break;
            }
        }
        fileOff ??= (int)thunkVa; // header area fallback (base assumed 0x10000 == file 0)

        int start = fileOff.Value;
        int off = start;
        while (off + 4 <= data.Length)
        {
            uint val = BinaryPrimitives.ReadUInt32LittleEndian(data.AsSpan(off, 4));
            if (val == 0) break;
            if ((val & 0x80000000) != 0)
            {
                int ordinal = (int)(val & 0x7FFFFFFF);
                imports.Add(new XbeKernelImport
                {
                    Ordinal = ordinal,
                    Name = XboxKernelExports.Name(ordinal),
                    ThunkAddress = (uint)(thunkVa + (off - start)),
                });
            }
            off += 4;
        }
        return imports;
    }
}
